osu.Game.Rulesets.Osu/Objects/Slider.cs

```diff
@@ -255,16 +281,38 @@ protected void UpdateNestedSamples()
 
             foreach (var tick in NestedHitObjects.OfType<SliderTick>())
                 tick.Samples = sampleList;
+        }
 
+        private void nodeSamplesOnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+        {
             foreach (var repeat in NestedHitObjects.OfType<SliderRepeat>())
-                repeat.Samples = this.GetNodeSamples(repeat.RepeatIndex + 1);
+            {
+                unbindOld(repeat.SamplesBindable, repeat.RepeatIndex + 1);
+                bindNew(repeat.SamplesBindable, repeat.RepeatIndex + 1);
+            }
 
-            if (HeadCircle != null)
-                HeadCircle.Samples = this.GetNodeSamples(0);
+            unbindOld(HeadCircle?.SamplesBindable, 0);
+            bindNew(HeadCircle?.SamplesBindable, 0);
+            unbindOld(TailSamples, repeatCount + 1);
+            bindNew(TailSamples, repeatCount + 1);
 
-            // The samples should be attached to the slider tail, however this can only be done after LegacyLastTick is removed otherwise they would play earlier than they're intended to.
-            // For now, the samples are played by the slider itself at the correct end time.
-            TailSamples = this.GetNodeSamples(repeatCount + 1);
+            void bindNew(BindableList<HitSampleInfo> bindableList, int index)
+            {
+                if (bindableList is null || e.NewItems is null || index < e.NewStartingIndex || index >= e.NewStartingIndex + e.NewItems.Count) return;
+
+                var newSamples = (BindableList<HitSampleInfo>)e.NewItems[index - e.NewStartingIndex];
+                bindableList.UnbindFrom(SamplesBindable);
+                bindableList.BindTo(newSamples);
+            }
+
+            void unbindOld(BindableList<HitSampleInfo> bindableList, int index)
+            {
+                if (bindableList is null || e.OldItems is null || index < e.OldStartingIndex || index >= e.OldStartingIndex + e.OldItems.Count) return;
+
+                var oldSamples = (BindableList<HitSampleInfo>)e.OldItems[index - e.OldStartingIndex];
+                bindableList.UnbindFrom(oldSamples);
+                bindableList.BindTo(SamplesBindable);
+            }
```

Comment:
I feel like this even proves my point above already. Is this how much code every usage of this double-nested-bindable-list is going to have to go with for it to work correctly?

Reply:
This code is more complex than a typical use-case of a double-nested-bindable-list, because it also has the `Headcircle` and `TailSamples` to bind with the node samples and there is this extra complexity of the `HitObject` samples being used as a default. 

I think usually you'd just add your binding on each added item and remove your binding on each removed item, so 2 for-loops.